File: server/makc2022--dotnet/Makc2022.Layer4.Sql.Domains.DummyMain/Queries/Item/Get/DomainItemGetQueryInput.cs
// Copyright (c) 2022 Maxim Kuzmin. All rights reserved. Licensed under the MIT License.

using Makc2022.Layer2.Sql.Queries.Item.Get;

namespace Makc2022.Layer4.Sql.Domains.DummyMain.Queries.Item.Get
{
    /// <summary>
    /// Входные данные запроса на получение элемента в домене.
    /// </summary>
    public class DomainItemGetQueryInput : ItemGetQueryInput
    {
        #region Properties

        /// <summary>
        /// Имя сущности.
        /// </summary>
        public string? EntityName { get; set; }

        #endregion Properties

        #region Public methods

        /// <inheritdoc/>
        public sealed override void Normalize()
        {
            base.Normalize();

            if (EntityId > 0)
            {
                EntityName = null;
            }
        }

        /// <inheritdoc/>
        public sealed override List<string> GetInvalidProperties()
        {
            var result = base.GetInvalidProperties();

            if (result.Any())
            {
                if (EntityName != null)
                {
                    result.Clear();
                }
                else
                {
                    result.Add(nameof(EntityName));
                }
            }

            return result;
        }

        #endregion Public methods
    }
}

File: server/makc2022--nodejs--nestjs/src/layer4/domains/dummy-one-to-many/queries/item/get/domain-item-get-query-input.ts
/** Copyright (c) 2022 Maxim Kuzmin. All rights reserved. Licensed under the MIT License. */

import { nameof } from '../../../../../../layer1/common/common-functions';
import { ItemGetQueryInput } from '../../../../../../layer2/nosql-mongo/queries/item/get/item-get-query-input';

export class DomainItemGetQueryInput extends ItemGetQueryInput {
  /** Имя сущности. */
  entityName: string;

  /** @inheritdoc */
  override normalize(): void {
    super.normalize();

    if (this.entityId) {
      this.entityName = null;
    }
  }

  /** @inheritdoc */
  override getInvalidProperties(): string[] {
    const result = super.getInvalidProperties();

    if (result.length > 0) {
      if (this.entityName !== null) {
        result.length = 0;
      } else {
        result.push(nameof<DomainItemGetQueryInput>('entityName'));
      }

      return result;
    }
  }
}
